### DISCUSSION TITLE
dealing with nulls and zero

### DISCUSSION THREAD

**Discussion by llfletch** [dotnet/csharplang#2532] [Q&A] [answered]
There seems to be a problem with how this language deals with nulls and zeros.
I will try to clear this up by explaining what they mean and how they should be used.
Null basically means not applicable (N/A).  If we have a field for paint cans, a null value would represent that we do not carry paint cans.  So how many paint cans do we have?  We have zero paint cans (we don't carry them).  Whenever we have a null value, it means we have zero.  A zero, however does not mean the value is null.  We could sell paint cans if we had them, but there aren't any at the moment.
Logical values are special and they potentially have one of four values: null, zero, true, and false. If we were to ask if a certain word is usually a noun, the result could be stored in a logical value. If we said "true" then we believe that the word is usually a noun, if we said "false", then we believe that it is usually not a noun.  "zero" would mean that we haven't decided and "null" would mean that the question wasn't relevant (and by definition that we haven't decided, since nulls are zeros also).
All variables can be converted to truth values.  All variables potentially have a "zero" value and a "true" value.  For numbers, positive numbers are "true", negative numbers are "false", and 0 is "zero".  The "zero" value means that there is no valid data and the "true" value means there is valid data.  For strings, empty strings and null strings, should have a "zero" truth value, otherwise they are "true".
Zero is what most variables should be initialized to.

Now that I have outlined the theory of zero, why is it relevant?
I will focus on the ? operator.  Lets look at the following statement:

`if (Employee?.Name == 0)`

Obviously, if there is valid data for Employee.Name, the statement should be false.
If Employee.Name is "" (empty string) it should be true.
What if Employee is null?
A null value is a zero as is an empty string, so the statement should be true in that case.
Dealing with nulls and zeros appropriately should be able to fix some of the main problems with the ? operator.

I would like to get feedback on what other problems this theory might address.

**Comment by YairHalberstadt** [comment, marked as answer]
The meaning of null, zero, empty string, etc. are all highly dependent on the context.

As such unlike C++ and JS, C# requires you to be explicit about how you want to treat them.

Given the sheer scale of the disaster type coercion is in JS this is definitely the right decision. Consider the fact that the `==` operator is now essentially banned from many JS codebases, and all usages are replaced with `===` which has a much stricter definition of equality.

**Comment by llfletch** [comment]
Would there be a way to indicate that you want it to act this way, but have it act normally otherwise?

**Comment by YairHalberstadt** [comment]
Yes
```csharp
if(string.IsNullOrEmpty(Employee?.Name))
```

**Comment by llfletch** [comment]
You say that the definitions of null, zero, empty strings are highly dependent on the context, It seems to me that people should "sit down" and figure out exactly what they should mean so that there can be some sort of standard.  Having things like that being "highly dependent on the context" is unacceptable.  It basically means that we have no idea what it means except by how it happens to be used.

**Comment by YairHalberstadt** [comment]
We know exactly what an empty string means. It means that a string contains no characters.

What's highly dependent on context is what you do with an empty string.

**Comment by llfletch** [comment]
Then we need to figure out how they should be used.

**Comment by YairHalberstadt** [comment]
They will be used differently depending on your application, and what the string represents.

**Comment by YairHalberstadt** [comment]
A string can represent a name, a collection of characters, a piece of text, a byte array etc. All of these cases will need to deal with empty strings differently.

An int can represent an age, the number of items in a collection, an Id, an offset, a port, or a thousand other things. Every single one of those scenarios treats 0 differently.

**Comment by llfletch** [comment]
And yet zero should mean the same thing in almost every case.  In almost every case it should be what the value is initialized to and it should mean that there is no valid data.

**Comment by llfletch** [comment]
One can always say things depend on the situation, but there is value in finding commonalities and underlying patterns.

**Comment by YairHalberstadt** [comment]
Is 0 a valid age?

Is 0 a valid network port?

Is 0 a valid price?

Is 0 a valid weight?

The answer to every single one of those questions is different:

Yes.
Yes but it's special cased.
In some systems yes, in others no, depending on how you represent a free item.
Not if you're measuring something physical, but possibly yes if you're rounding, or measuring something non-physical.

**Comment by yaakov-h** [comment]
I'm gonna re-post something I put in another GitHub Issue here:

`null` is an abomination that should never have existed.

`null` is a lie that people ignore.

`null` is the computer saying "hey I have that data you want, it's at memory location zero" and all code checking for it saying "yeah, right, I don't believe you, if I go looking at memory location zero I will get my head bitten off."

Because, if it does, it will hit the first page of memory, which is reserved by the operating system, and trigger a page fault. In the CLR, this becomes a `NullReferenceException`, and in other runtimes you can just crash.

When it comes to most data structures, `null` is a legitimate nothing. There is nothing there. Not an empty thing, a nothing.

---

>  If we have a field for paint cans, a null value would represent that we do not carry paint cans.

If we have a field for paint cans, an _empty array or empty Enumerable_ should represent that we do not carry paint cans. `null` doesn't mean that we have paint cans, or that we don't have paint cans, rather we have only laid a trap for the caller.

If it's a field for a singular paint can, then null could be indicative of no paint can, and this is it's generally accepted use. This does not mean zero.

> A zero, however does not mean the value is null. We could sell paint cans if we had them, but there aren't any at the moment.

So if we have none, it could be zero, or it could be null, and if it's null then there are zero, but if there are zero it might not be null?

Using two different values interchangeably to mean the same thing is a 'fun' vector for bugs.

> Logical values are special and they potentially have one of four values: null, zero, true, and false. If we were to ask if a certain word is usually a noun, the result could be stored in a logical value. If we said "true" then we believe that the word is usually a noun, if we said "false", then we believe that it is usually not a noun. "zero" would mean that we haven't decided and "null" would mean that the question wasn't relevant (and by definition that we haven't decided, since nulls are zeros also).

This sounds like you need an enum. There is no current type in the BCL that can reflect these four states.

This also reminds me of the [tri-state bool](https://thedailywtf.com/articles/What_Is_Truth_0x3f_). 

> All variables can be converted to truth values. All variables potentially have a "zero" value and a "true" value. For numbers, positive numbers are "true", negative numbers are "false", and 0 is "zero". The "zero" value means that there is no valid data and the "true" value means there is valid data.

So what are negative numbers then, if we've already used "zero" for invalid and "true" for valid?

Assigning roles like this may make sense in some limited domains, but this isn't something you can generalize to the language level.

>  For strings, empty strings and null strings, should have a "zero" truth value, otherwise they are "true".

"null strings" are not a thing. `null` has no type. If it's `null` then it's not a string; if it's a string then it's not `null`.

A lot of the standard library treats `null` as empty string when performing concatenation and formatting operations, but this does not mean that they are the same thing.

> Zero is what most variables should be initialized to.

All variables are initialized to zero by default, and for reference types, that's a zero pointer, also known as `null`. An empty string is not a "zero".

**Comment by HaloFour** [comment]
@llfletch 

That C# makes you very explicitly deal with the differences in data types is not a problem.  It is very intentional and based on the fact that languages that permit such implicit coercion and "truthiness" are loaded with bug vectors.

`null` is not 0 is not an empty string, nor should any of those things be confused with one another.  C# is very intentionally trying to not be JavaScript.

**Comment by llfletch** [comment]
Maybe the solution would be to allow a ~= operator that could make such coercion possible without sacrificing the type strictness of the nominal case.

**Comment by DavidArno** [comment]
@llfletch,

As others have said, what you are proposing leads to some of the worst bugs in JavaScript, leading to many to ban the use of `==` in that language. Are you therefore able to supply a real-world example of where bypassing the protection from these bugs that C# offers us would be useful?

**Comment by llfletch** [comment]
That is a good question.  First of all, the people using the ~= (equivalency) operator would have to know what they were getting into, the idea is to provide convenience for people who know the risks associated with such loose typing.  Mostly I think it could be useful as a shorthand for things that would normally require a type cast or a more lengthy solution.

An example:
`Employee?.Name ~= ""`

In this example, if Employee is null, the == operator would return false, but since null and empty string are roughly equivalent (they both represent an absence of data), the ~= operator would return true.

I realize that we could use the following (as stated earlier):
`if (string.IsNullOrEmpty(Employee?.Name))`
The problem with this is that it is lengthy and depends on the user to know that there is such a function.
The equivalency operator would be much easier to remember and could apply to a number of other uses.

Since the previous example may be a poor example, let me give another example:
`object1 ~= true`
Now object1 could be any object, but it would be "true" if it had any values that were different from their initial state (presumably zero).  So the object being equivalent to true would mean that we had data and false would mean that we didn't have data in it.

**Comment by YairHalberstadt** [comment]
How do you define roughly equal?

**Comment by llfletch** [comment]
That is a good question that i don't have an answer to.  Obviously it would take some discussions and planning and thought about what "roughly equal" would mean.  I don't presume to know exactly, I am just proposing that we make a standard as to what that would mean.

**Comment by HaloFour** [comment]
@llfletch 

> The problem with this is that it is lengthy and depends on the user to know that there is such a function.

That function very clearly specifies to both the developer and anyone maintaining that code what it does and what values it would consider to be empty.  This feature requires that the user be aware of how this "rough equivalence" (aka "false equivalence") works and all of the myriad of ways it's likely to lead to unexpected logic errors.  Best of all, functions like that require zero language changes.

`""` and `null` are not remotely equivalent and no language feature should implicitly consider them as such.  What's next, all white space?  Empty strings and empty arrays?  You don't have to dig deep into the problems of "truthiness" in Javascript to see how problematic it becomes, and quickly.

**Comment by llfletch** [comment]
"" and null are remotely equivalent precisely because they represent an absence of data.  Empty strings and empty arrays are remotely equivalent for the same reason.  White space, however, is data.  Yes there are problems in Javascript, but this is because they went about it wrong and didn't set strict enough standards and meanings for things.

**Comment by HaloFour** [comment]
The lack of stuff in a container and the lack of a container are not equivalent.

If you want to consider the various states of a type to be equivalent, you can write a function.  Adding a language feature that violates the design principals of C# seems like a non-starter.

**Comment by llfletch** [comment]
If a function would be better for this purpose, then how about a standard function with a declaration like the following:
bool IsData()
It could be overloaded to handle all of the different data types and could address part of the same issue in a different way.  The main thing is that it is available to everyone.

**Comment by HaloFour** [comment]
> then how about a standard function with a declaration like the following:
> bool IsData()

It has a meaningless name that doesn't tell you what it does.  But feel free to write one and add it to your project.

**Comment by llfletch** [comment]
Yes I could write one and add it to my project, I could even define a new operator  for myself.  The whole point of me posting my ideas here is to help advance the progress of the C sharp language for everyone.

**Comment by CyrusNajmabadi** [comment]
>  The whole point of me posting my ideas here is to help advance the progress of the C sharp language for everyone.

From C#'s perspective (as well as the community and ecosystem), it's very much a *good* thing that a null string is not an empty string is not the number 0.  These are distinct concepts.  For any particular domain you use whatever concepts are appropraite for you, such as:

```
s == null
s == ""
string.IsWhitespace(s)
string.IsNullOrWhitespace(s)
IsPhoneNumber(s)
regex.Matches(s)
```

etc.   There's no universal concept of waht is or isn't valid data.  So you build the right abstractions for your particular domain.

**Comment by llfletch** [comment]
It is absolutely true that null is different from 0 which is different from empty string.  It is true, these are definitely distinct concepts.
Why is it so bad to have a new concept that there is either meaningful data in a variable or there is not meaningful data in a variable (whatever the variable may be).
Is it not possible to make such a concept?

**Comment by HaloFour** [comment]
@llfletch 

> Why is it so bad to have a new concept that there is either meaningful data in a variable or there is not meaningful data in a variable

We already have that concept.  We can compare a variable to `null` or to `0` or to `default` or to whatever you consider to be not "meaningful data".  The language doesn't have or need a way to do this across disparate data types or to consider `0` to be equivalent to `null`, `""`, `[]` or any other combination of values.

**Comment by CyrusNajmabadi** [comment]
> Why is it so bad to have a new concept that there is either meaningful data in a variable or there is not meaningful data in a variable (whatever the variable may be).

How do you define what is meaningful?  As i showed above it's domain specific.  For some domains a string isn't meaningful if it's full of whitespace, in others it would be.

> Is it not possible to make such a concept?

Not really no.  I recommend you try as it might be very informative.

**Comment by llfletch** [comment]
We do not have that concept.  Remember, null, 0 and "" are different and distinct concepts.  I disagree, the language does need such a concept and badly because if we don't have this concept, we end up compartmentalizing everything and we end up not being able to generalize to a higher level.  By not being able to generalize to higher level concepts, we end up being stuck at a lower level of detail.  One of the main problems with computing is having to do everything at a low level of detail and not being able to handle higher level abstractions.  This slows down software development.

**Comment by HaloFour** [comment]
> Remember, null, 0 and "" are different and distinct concepts.

Yes, they are, and they should be.

If you want to treat them as equivalent, write a function to do so.

> I disagree, the language does need such a concept and badly because if we don't have this concept, we end up compartmentalizing everything and we end up not being able to generalize to a higher level.

I disagree.  Generalizing this concept and across data types is one of the leading causes of bugs in the languages that allow it.  C# made the decision to not treat `null` like `0` like `false` specifically because of learning from the bad experiences from C and C++.  Truthy/Falsy is a constant source of bugs in JavaScript.

If you want this generalization, write your own function.

> This slows down software development.

I'll take slightly slower software development resulting from having to explicitly consider my data types over chasing down the inevitable flood of bugs any day of the week.

**Comment by llfletch** [comment]
How do I define what is meaningful?  It seems like it is domain specific, but it could be generalized possibly with some user preferences.  White space is meaningful in general, but there could always be a preference or option to change this for a particular user or circumstance.  Just because there are problems with the concept doesn't mean that it couldn't be done - if we allow it to be flexible.

**Comment by spydacarnage** [comment]
Remember, C# 8 is doing a lot to deal with the null problem - it's moving
towards getting rid of them!  The last thing we need is a new feature that
makes them more prevalent again...

**Comment by CyrusNajmabadi** [comment]
> > We do not have that concept. Remember, null, 0 and "" are different and distinct concepts.
>
> I disagree, the language does need such a concept and badly because if we don't have this concept, we end up compartmentalizing everything and we end up not being able to generalize to a higher level. By not being able to generalize to higher level concepts, we end up being stuck at a lower level of detail. One of the main problems with computing is having to do everything at a low level of detail and not being able to handle higher level abstractions. This slows down software development.

What is your concept?  Please explain it fully.

You have also avoided the points i've made.  Such as the fact that whether or not data is meaningful is unrelated to whether or not it is null, empty or 0.  For example, the empty string is *very* meaningful in many domains.  But it's completely non-meaningful in others.  How do you create your concept such that it deals with this fact?

**Comment by CyrusNajmabadi** [comment]
> White space is meaningful in general, but there could always be a preference or option to change this for a particular user or circumstance. 

Yes.  We have a way to specify that preference.  But using functions like `string.IsWhitespace(...)`.  How would you encode all the preferences for strings?  What if printable characters are meaningful for one person, but not for another?  What if special characters are meaningful for one, but not another.  What if only US phone numbers are meaningful for one, but not hte other?  What if all phone numbers are meaningful for one but not the other?

etc. etc. etc.

How are you actually going to provide a way for people to specify hte infinite number of meanings of what is "meaningful" or not?

**Comment by llfletch** [comment]
I will try to clarify what I meant be meaningful data.
For strings, null or "" would not be meaningful, but "Jason", or "Network" would be.  Obviously if we have just white space that is a problem that could be addressed with preferences, But if there are any letters, numbers or special characters, they are meaningful (even if they are incorrect or just noise).
For boolean types, false or zero or null (if we have that option) is not meaningful by default, but true values are.  There is a problem with this because of the overloading of meaning of false that we want to believe that the question is not true or that we don't know. The default should be that only true is meaningful in this concept with perhaps some options to specify how they want to deal with this if they want to deal with it in another way. 
For number types, positive numbers should always be meaningful and zero should not be meaningful, and negative numbers should default to being meaningful.
For arrays, if any of the array elements are meaningful then the whole array is meaningful.
For variables based on enum types, if the type is null (if that is possible) or can be converted to a zero number as is the case with the first element normally the enum type should by default be not meaningful, otherwise it is meaningful.  I realize this may not be the behavior every one wants, but that should be the default.
Yes there are a whole slew of preferences that could be considered and would complicate things greatly if they were implemented, but the idea is to keep things relatively simple so that most thing act the way we would want and if they don't they can always be addressed the way they are now.

**Comment by CyrusNajmabadi** [comment]
> `""` would not be meaningful

But empty strings are *very* meaningful in many domains.

> For boolean types, false or zero or null (if we have that option) is not meaningful by default

boolean types *only* have `true` or `false`.  Why would it be a good idea to add `zero` or `null` into that mix?

> For number types, positive numbers should always be meaningful and zero should not be meaningful

Why is zero not meaningful?  What about all the domains where 0 is a perfectly expected and natural value?

--

Regardless, here's your solution:

```c#
public static bool IsMeaningful(this object o)
    {
        switch (o)
        {
            case string s: return s.Length > 0;
            case bool b: return b;
            case null: return false;
            case Array arr: return arr.OfType<object>().Any(a => a.IsMeaningful());
            default:
                try { return Convert.ToInt64(o) != 0; } catch { }
                try { return Convert.ToUInt64(o) != 0; } catch { }
                return false;
        };
    }
```

You don't need a language helper for this.  You can just define your helper and put it on nuget and people can use it if they want.

**Comment by CyrusNajmabadi** [comment]
> so that most thing act the way we would want 

Here's teh crux of the problem: "the way we would want" is person, team, group and domain dependent.

I absolutely do think of empty strings as meaningful in my domains.  0 is also meaningful.  Arrays have no concept of meaningfulness for me whatsoever.  etc. etc.  

You keep saying this as if there's consensus in the c# ecosystem about this, but there isn't.  And there's broad belief that languages that *have* done similar things (i.e. JavaScript) screwed up here as this is *not* a good concept to have in the language.  C# very intentionally and explicitly has not done this because it's actively felt to be a bad enough idea by enough of the LDM and enough of the community.

**Comment by llfletch** [comment]
It is a shame no one can come to a consensus. I believe the idea is good, but obviously the people that have tried similar things like JavaScript have failed to get it right and have spoiled people's opinion of trying such things.  Obviously you are right "the way we want" is person, team, group and domain dependent, but there is also common threads or ideas that could make for a new concept or standard if people were open to it.  Thanks everyone for listening to this failed idea because it has helped me to understand the problems we face.

**Comment by CyrusNajmabadi** [comment]
Good luck with your journey!  As i mentioned above @llfletch: i highly recommend you actually try this out and create a library you think woudl be valuable here.  You can even host that library on nuget, and if it's positively viewed by the community, will get used in more and more places.  If it turns out that you, along with the community, can form a good set of rules here, it's possible in the future that thoughts on this might be different.

**Comment by llfletch** [comment]
That is exactly what I plan to do. Thank you.



Sent from my Verizon, Samsung Galaxy smartphone


-------- Original message --------

**Comment by DavidArno** [comment]
@llfletch,

> For number types, positive numbers should always be meaningful and zero should not be meaningful...

You have a goal of trying to define a common set of principles around equivalence between different values. But that is not possible as there is no common set of principles. There literally can't be because domains vary so much.

And there's no better example of this than the exit status from commands in Linux. Zero means success. Non-zero means there was an error. This even embodied in two bash commands, `true` (returns 0) and `false` (returns 1, as an arbitrary non-zero value). So the moment my C# app running on Linux collides with your  ideas on what is and isn't meaningful and what is and isn't equivalent, all hell breaks loose.

Zero absolutely is meaningful, sometimes. And that's the problem: for any rule you invent, sometimes that rule will be plain wrong for a given domain. `""` is data sometimes; sometimes it's an absence of data. The same applies to `0`, `null`, whitespace, comments, `-1` etc etc. I might have a system where 0-99 are valid values, and so -n and 100+ all mean "no value". Do we really want some configurable system of "roughly equal to" that would therefore treat `100 ~= false` as a true statement? And if it's not configurable, what use is it to anyone that strays from some completely arbitrary set of rules on rough equivalence?
